pkg/adapter/config/handler.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"io"
+
+	"github.com/golang/protobuf/proto"
+)
+
+type (
+	// Handler represents default functionality every Adapter must implement.
+	Handler interface {
+		io.Closer
+	}
+
+	// AdapterConfigValidator handles adapter configuration defaults and validation.
+	AdapterConfigValidator interface {
+		// DefaultConfig returns a default configuration struct for this
+		// adapter. This will be used by the configuration system to establish
+		// the shape of the block of configuration state passed to the Configure method.
+		DefaultConfig() proto.Message
+		// ValidateConfig determines whether the given configuration meets all correctness requirements.
+		ValidateConfig(proto.Message) error
+	}
+
+	// HandlerBuilder represents a factory of Handler. Adapters register builders with Mixer
+	// in order to allow Mixer to instantiate Handler on demand.
+	HandlerBuilder interface {
+		io.Closer
+		AdapterConfigValidator
+
+		// Name returns the official name of the adapter produced by this builder.
+		Name() string
+		// Description returns a user-friendly description of the adapter produced by this builder.
+		Description() string
+		// ConfigureHandler is called by Mixer to pass the Handler specific configuration to the Adapter implementation.
+		ConfigureHandler(cnfg proto.Message) error
```

Comment:
Why is `ConfigureHandler` separate from `Build`? I think we want:
```go
Build(cfg proto.Message) (Handler, error)
```

I think this is essentially the same problem we discussed on noop2 above: we want creation and configuration to be a single atomic thing.